src/java_bytecode/generate_java_generic_method.cpp

```diff
@@ -0,0 +1,296 @@
+/*******************************************************************\
+
+Module: Generate Java Generic Method - Instantiate a generic method
+        with concrete type information.
+
+Author: DiffBlue Limited. All rights reserved.
+
+\*******************************************************************/
+
+#include "generate_java_generic_method.h"
+#include <java_bytecode/java_types.h>
+#include <util/expr_iterator.h>
+
+generate_java_generic_methodt::generate_java_generic_methodt(
+  message_handlert &message_handler):
+    message_handler(message_handler)
+{}
+
+/// Generate a copy of a given generic method, specialized with
+/// the given concrete types and insert the method into the symbol
+/// table.
+/// \param generic_method The generic method to be specialized.
+/// \param concrete_types A map from generic type variables to concrete types
+/// with which to instantiate the generic method.
+/// \param symbol_table The symbol table into which the generated method will
+/// be inserted.
+/// \return The symbol that was inserting into the symbol_table, or the existing
+/// symbol if the method has already been specialized.
+const symbolt& generate_java_generic_methodt::operator()(
+  const symbolt &generic_method,
+  const type_variable_instantiationst &concrete_types,
+  symbol_tablet &symbol_table) const
+{
+  INVARIANT(generic_method.type.id()==ID_code, "Only code symbols");
+  INVARIANT(concrete_types.size()>0, "Should be at least one concrete type");
+  // Another invariant is that concrete_types.size() > the total number of
+  // type variables used in the method. However, determining that essentially
+  // means iterating through the whole method structure and duplicating a
+  // large chunk of the work we're about to do...
+
+
+  // We need to decorate the signature of the specialized method to
+  // differentiate the symbols of the generic method and the specialized
+  // method. There's a design decision to be made here - do we actually modify
+  // the java signature, or do we decorate it with additional
+  // qualifiers. We are taking the approach of adding decorations because
+  // that avoids the situation where the user may have already
+  // declared an overloaded method that might have the same signature
+  // as the specialized method we are about to create. By decorating the
+  // existing symbol, we ensure that no legal Java program could have already
+  // defined the symbol.
+  const std::string& signature_decoration=instantiation_decoration
+    (concrete_types);
+
+  // Check the method has not already been specialized
+  symbol_tablet::symbolst::const_iterator already_specialized=
+    symbol_table.symbols.find(
+      id2string(generic_method.name)+signature_decoration);
+  if(already_specialized!=symbol_table.symbols.end())
+    return already_specialized->second;
+
+  // Convert concrete types into a map for faster lookups
+  type_variable_instantiation_mapt concrete_type_map(concrete_types);
```

Comment:
Two questions (that may become clear as I keep reading):
1. why not just accept as input a map?
2. faster lookup seems unlikely to have a big impact when  I've never seen a class with more than 3 but I also suppose this will be more readable anway...

Reply:
The problem with just accepting a Map as the input is the lack of ordering. At the moment we decorate the method identifier with a list of the concrete types *in the order they are provided in the initlializer_list* - this means that we get a consistent ordering of types in the identifier decoration. e.g. if we have 
```java
class Foo <A, B> {
void someMethod(A a, B a) {...}
}
```
then we pass in an intializer_list that looks something like `{{A,Integer},{B,Double}}` - which then gets used to decorate the method identifier with '<Integer, Double>'  which matches the ordering of type parameters in the class definition. However, if we just passed in a `Map<typevar,concretetype>` - we don't know whether we'd end up with '<Integer,Double>' or '<Double,Integer>', depending on the implementation of Map... 

Of course, perhaps the correct answer is to use a Map, but instead of constructing the decoration string from the Map, go look up the parent class (or type parameters for generic methods in non-generic classes) to find the ordering of the type parameters.